Documentation Update : resource-quota.md (Follow-up to PRs #1160 and #1162) (#1183)
netgroup-polito/CrownLabs: operators/docs/resource-quota.md

# Defining Quotas per Workspace and per User

This document outlines the architecture, data models, enforcement mechanisms, and instructions for managing resource quotas within CrownLabs.

---

## 1. Defining Quotas per Workspace

Each workspace defines its maximum allowed resource allocation in the `spec.quota` field of the CrownLabs `Workspace` Custom Resource. 

Example configuration for the `test` workspace:

```yaml
apiVersion: crownlabs.polito.it/v1alpha1
kind: Workspace
metadata:
  name: test
spec:
  prettyName: CrownLabs workspace dedicated to testing
  quota:
    cpu: 10
    instances: 3
    memory: 64Gi
    disk: 15Gi
    otherResources:
      nvidia.com/gpu: "1"
      amd.com/gpu: "1"
```

> In the example above, users belonging to this workspace can launch up to **3 active instances** (VMs or containers) simultaneously, and globally across the workspace cannot exceed **10 CPUs**, **64Gi of RAM**, **15Gi of Persistent Disk**, **1 NVIDIA GPU**, and **1 AMD GPU**.

Reviewer: ```suggestion
In the example above, users belonging to this workspace can launch up to **3 active instances** (VMs or containers) simultaneously, and cannot exceed **10 CPUs**, **64Gi of RAM**, **15Gi of Persistent Disk**, **1 NVIDIA GPU**, and **1 AMD GPU** globally within the workspace .
```

> Resource quotas can be updated at any time by modifying the `quota` section of the `Workspace` resource.

Reviewer: ```suggestion
Resource quotas can be updated at any time by modifying the `quota` section of the `Workspace` resource.
In case resources are decreased, no enforcement is done for all the instances that are already running; hence, users who have already running instances may (temporarily) consume more resources than the maximum allowed values within the workspace.
```

### Resource Types Specification

The data types used within `ResourceSpec` in the Go backend are defined as follows:

| Resource Field | Go Data Type | Description |
| :--- | :--- | :--- |
| `instances` | `int64` | Maximum number of active/running instances allowed |
| `cpu` | `int64` | Maximum vCPU cores allocated |
| `memory` | `resource.Quantity` | Maximum RAM memory (e.g., `16Gi`) |
| `disk` | `resource.Quantity` | Maximum persistent storage allocated (e.g., `20Gi`) |
| `otherResources` | `map[string]resource.Quantity` | Generic key-value map for extended hardware accelerators |

> ---

Reviewer: ```suggestion
---
`Resource.Quantity` is a common format used in Kubernetes, consisting in a number and a suffix (e.g., 16Gi).
More information [here](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/resource#Quantity).
```

For instance, if a user is enrolled in:
* **Workspace A:** CPU limit = `10`
* **Workspace B:** CPU limit = `8`

The total accumulated CPU limit for this user across the platform is `18`. 

> **NOTE on CPU Limits:** CPU cores are summed across workspaces. However, individual `Templates` (which belong to a specific Workspace) can define an extra property called `ReservedCPUPercentage` (e.g., `50%`), which throttles the actual vCPU allocation of instances spawned from that template.

### Personal Workspace Quota

Beyond the accumulated limits inherited from course workspaces, each user is allocated a **Personal Workspace**. This acts as a private, standalone sandbox. Its quota is configured directly within the `personalWorkspace` field of the `Tenant` Custom Resource:

```yaml
apiVersion: crownlabs.polito.it/v1alpha2
kind: Tenant
metadata:
  name: s343940
spec:
  firstName: John
  lastName: Doe
  personalWorkspace:
    cpu: 4
    memory: 8Gi
    instances: 2
    disk: 10Gi
    otherResources:
      nvidia.com/gpu: "0"
```

---

## 3. How to Add New Custom Resources (Step-by-Step Guide)

> CrownLabs uses `otherResources: map[string]resource.Quantity` to avoid hardcoding specific hardware vendor keys in the backend Go codebase. Follow these steps to introduce a new custom resource (e.g., **Intel Gaudi TPU** `intel.com/gaudi` or new **AMD GPUs** `amd.com/gpu`):

Reviewer: ```suggestion
While currently the mostly used _other resource_ are Nvidia GPUs, we cannot exclude that in the future we may need to define new custom resources, such as **Intel Gaudi TPU** `intel.com/gaudi` or new **AMD GPUs** `amd.com/gpu`.

CrownLabs source code has been created in order to support new resources without touching the code itself; in fact, it uses `otherResources: map[string]resource.Quantity` to avoid hardcoding specific hardware vendor keys in the backend Go codebase.

This section documents the steps required to add a new custom resource within CrownLabs, which will be automatically applied in both the frontend (so that workspace administrator can set their limits) and in the backend (by the code that implements resource quota enforcement).
```

```yaml
otherResources:
  intel.com/gaudi: "1"
```

## 4. Enforcement Mechanism: ResourceQuota vs Validation Webhook

CrownLabs enforces resource limits using a multi-tiered security and validation pipeline:

```
[ User Request ] 
       │
       ▼
[ 1. Validating Webhook ] ──(intercepts via namespaceSelector)──► [ Rejects if Invalid / Over-quota ]
       │
       ▼
[ 2. Instance Validator ] ──(Go component)─────────────────────► [ Verifies Workspace bounds ]
       │
       ▼
[ 3. ResourceQuota ]      ──(Tenant Namespace Level)───────────► [ Passive Global Security Barrier ]
```

1. **Validating Webhook (Cluster Level Enforcement):**
   CrownLabs deploys a `ValidatingWebhookConfiguration` that intercepts `CREATE` and `UPDATE` operations on `Tenants` and `Instances`. 
   * **Scope Filtering:** Uses `namespaceSelector` to match all CrownLabs-managed namespaces (`workspace-*` and `tenant-*`). This ensures **100% of instance creations are intercepted**, preventing validation bypasses.
   * **Failure Policy:** Configured with `failurePolicy: Fail` to block unvalidated or over-quota requests before they are written to `etcd`.

> 2. **ResourceQuota (Tenant Namespace Barrier):**

Reviewer: The ResourceQuota is not clear to me.
1. Which values are set for this resource quota?
2. What's the reason for this mechanism?

Author: 1. The ResourceQuota generated inside a user's namespace is configured with the "aggregated total allowance" of that specific user. This value equals the sum of the quotas across all Workspaces they are enrolled in, plus their Personal Workspace quota (e.g., `hard.cpu = sum(workspace_cpus) + personal_cpu`, `hard.memory = ...`).
2. It serves as a safety net at the namespace boundary. While the custom Webhook and Operators handle CrownLabs-specific business logic at the Custom Resource level (`Tenant`, `Instance`), the native `ResourceQuota` operates directly on low-level Kubernetes primitives (`Pods`, `PVCs`). It ensures that even if custom code experiences an unhandled failure or bypass, the Kubernetes API server natively blocks any pod or container from exceeding the user's hard limits.

   When a Tenant is reconciled, the Tenant Operator creates a standard Kubernetes `ResourceQuota` inside the user's namespace (`tenant-<username>`). It acts as a passive container limit preventing cluster starvation.

3. **Instance Validator (Workspace Level):**

>    A dedicated Go component that evaluates the real-time consumption of the user against their allowed course workspace limits before an instance is allowed to boot.

Reviewer: Not clear to me why we need to have these 3 cascading mechanisms.
Can you please clarify?

Author: The three mechanisms operate at different levels of the stack, though the validators work together just at different levels. ResourceQuota normally isn't used if the validators are on, things work like this :

1. Validating Webhook (API Server Gatekeeper): it executes Synchronous pre-checks before objects are written to `etcd`. This provides immediate feedback (e.g., HTTP 422 errors to GraphQL/UI) and evaluates complex CrownLabs business rules (e.g., verifying custom GPU labels, template-level CPU throttles, multi-workspace relationships) which standard Kubernetes primitives cannot understand. To simplify it, it's a check BEFORE the instance is created.

2. Instance Validator (Operator Reconciliation Loop): The difference with the webhook is that it acts at the runtime level. So it's an Asynchronous runtime evaluator inside the Instance Operator controller. It's necessary because it handles race conditions, dynamic runtime updates, and workspace-specific bounds (ex. checking if a specific course workspace quota is respected when launching an instance). It also manages recovery loops (ex. phase shifting to `ResourceQuotaExceeded` and requeuing every 60s during persistent storage allocation failures). To give you an example : if disk creation (PVC), which happens at runtime, fails then the instance validator catches that error and shifts the phase to ResourceQuotaExceeded, after 60 seconds it tries to allocate the disk again, this is a type of error that the webhook validator would not be able to catch since it's something that might happen at runtime.

3. ResourceQuota (fallback) : like i said in the previous comment reply, this is useful if the validator is off it applies some limits to a user if the validator is turned off instead of just leaving it to the native kubernetes limits. The reason why this check only works if the validators are off is because the validator checks are much more restrictive (validator limits resources based on the single workspace you want to create the instance in, ResourceQuota limits resources based on the sum total of all resource limits from all workspaces a user is in (including personal workspace).

if you need any other clarifications or if the document requires some modifications feel free to ask.
